lib/pbench/test/functional/server/test_put.py

```diff
@@ -0,0 +1,63 @@
+from pathlib import Path
+import time
+
+from pbench.client import PbenchServerClient
+
+TARBALL_DIR = Path("lib/pbench/test/functional/server/tarballs")
+
+
+class TestPut:
+    tarballs: list[Path] = []
+
+    def test_upload_all(self, server_client: PbenchServerClient, login_user):
+        """Upload each of the pregenerated tarballs, and perform some basic
+        sanity checks on the resulting server state.
+        """
+
+        for t in TARBALL_DIR.glob("*.tar.xz"):
+            self.tarballs.append(t)
+            response = server_client.upload(t)
+            assert response.ok
+        datasets = server_client.get_list(
+            metadata=["server.tarball-path", "server.status"]
+        )
+        found = sorted(d.name for d in datasets)
+        expected = sorted(t.name[:-7] for t in self.tarballs)
+        assert found == expected
+        for dataset in datasets:
+            assert dataset.name in dataset.metadata["server.tarball-path"]
+            assert dataset.metadata["server.status"]["upload"] == "ok"
+
+    def test_index_all(self, server_client: PbenchServerClient, login_user):
+        """Wait for datasets to reach the "Indexed" state, and ensure that the
+        state and metadata look good
+        """
+
+        start = time.time()
+        timeout = start + (60.0 * 5.0)
+        done = False
+
+        # Test get_list pagination: to avoid forcing a list, we'll count the
+        # iterations separately
+        count = 0
+        datasets = server_client.get_list(limit=5)
+        while not done:
+            for dataset in datasets:
+                count += 1
+                metadata = server_client.get_metadata(
+                    dataset.resource_id, ["dataset.state", "server.status"]
+                )
+                state = metadata["dataset.state"]
+                assert state in ["Uploaded", "Indexing", "Indexed"]
+                if state != "Indexed":
+                    assert (
+                        time.time() < timeout
+                    ), "Exceeded timeout waiting for indexing"
+                    time.sleep(30.0)  # sleep for half a minute
+                    break
```

Comment:
Is this looping structure set up this way intentionally?  That is, if we hit a dataset which is not yet indexed, we'll go back and recheck the previously-checked ones...which is...interesting....

What's even more interesting is that (I _think_...) in this case we end up counting the initial tarball multiple times...which leaves me wondering how we ever pass the assertion at line 63.

It seems like we want a loop around lines 47-57 which does the wait for `"Indexed"` without advancing the outer loops.  If _that_ loop times out, then the game is up; and, we should reset the timeout when we start each dataset (if we increase the overall number of datasets, it would be nice if we didn't have to adjust the timeout -- I think that all we really care about is that the server is making progress).  So, we don't actually need the outer `while` loop, unless you're really trying to count _all_ the datasets (in which case, I think you need to have the call to `get_list()` inside the outer loop!).

Reply:
The server stages operate asynchronously in parallel, so waiting for each dataset separately wouldn't really make a difference. Since `get_list` is a generator rather than a list, I don't think that "restarting" to `for dataset in datasets` loop really restarts anything; I think it's just the next yield in the sequence. In any case, clearly this logic isn't as clear as it could be.

Reply:
I misplaced the fact that `datasets` is a generator.  (See also [here](https://github.com/distributed-system-analysis/pbench/pull/3064#discussion_r1022961739).)